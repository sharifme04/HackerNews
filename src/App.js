import React, { useState, useEffect, Suspense } from "react";
import axios from "axios";
import "./App.css";
const Pagination = React.lazy(() => import("./components/Pagination"));
const StoryComponent = React.lazy(() => import("./components/StoryComponent"));

const API_END_POINT = process.env.REACT_APP_BASE_PATH;

function App() {
  const storiesPerPage = 28;
  const [stories, setStories] = useState([]);
  const [currentPage, setCurrentPages] = useState(1);

  useEffect(() => {
    fetchStories();
  }, []);

  const fetchStories = async () => {
    let items = [];
    await axios
      .get(`${API_END_POINT}/topstories.json?print=pretty`)
      .then(response => response.data)
      .then(data => {
        data.map(async storyId => {
          await axios
            .get(`${API_END_POINT}/item/${storyId}.json?print=pretty`)
            .then(response => response.data)
            .then(async itemDetail => {
              items = [...items, itemDetail];
              await setStories(items);
            });
        });
      });
  };

  const indexOfLastStory = currentPage * storiesPerPage;
  const indexOfFirstStory = indexOfLastStory - storiesPerPage;
  const currentStories = stories.slice(indexOfFirstStory, indexOfLastStory);

  const paginate = pageNumber => setCurrentPages(pageNumber);
  const previousPage = pageNumber => setCurrentPages(pageNumber - 1);
  const nextPage = pageNumber => setCurrentPages(pageNumber + 1);

  return (
    <div className="container-fluid">
      <h1 className=" mt-1">Hacker News App</h1>
      <hr />
      <Suspense fallback={<div className="spinner-border"></div>}>
        <Pagination
          storiesPerPage={storiesPerPage}
          totalStories={stories.length}
          paginate={paginate}
          nextPage={nextPage}
          previousPage={previousPage}
          currentPage={currentPage}
        />
        <StoryComponent stories={currentStories} />
      </Suspense>
    </div>
  );
}

export default App;
