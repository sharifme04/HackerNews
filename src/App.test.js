import React from 'react';
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import App from './App';

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});
it("renders story data", async () => {
  const fakeStory = {
    by: "jger15",
    descendants: 15,
    id: 22000126,
    kids:[22001188, 22001236, 22001202, 22001107, 22001179, 22001091, 22001154, 22001106],
    score: 38,
    time: 1578567242,
    title: "Work on these things",
    type: "story",
    url: "https://marginalrevolution.com/marginalrevolution/2019/12/work-on-these-things.html"
  };

  jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(fakeStory)
    })
  );

  // Use the asynchronous version of act to apply resolved promises
  await act(async () => {
    render(<App id="22000126" />, container);
  });

  global.fetch.mockRestore();
});
