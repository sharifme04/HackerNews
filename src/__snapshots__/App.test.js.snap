// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> snapshot renders 1`] = `
<div
  className="container-fluid"
>
  <h1
    className=" mt-1"
  >
    Hacker News App
  </h1>
  <hr />
  <div
    className="spinner-border"
  />
</div>
`;
